File: leetcode/Medium/2192. All Ancestors of a Node in a Directed Acyclic Graph/2192.java
class Solution {
    public List<List<Integer>> getAncestors(int n, int[][] edges) {
        List<List<Integer>> res = new ArrayList<>();
        Map<Integer, List<Integer>> map = new HashMap<>();

        for(int i=0; i<n; i++){
            map.put(i, new ArrayList<>());
            res.add(new ArrayList<>());
        }

        for(int[] e:edges){
            map.get(e[1]).add(e[0]);
        }

        for(int i=0; i<n; i++){
        	// dfs(i, i, map, res, new boolean[n]);
            bfs(i, map, res, new boolean[n]);
            Collections.sort(res.get(i));
        }
        return res;
    }

    private void bfs(int idx, Map<Integer, List<Integer>> map, List<List<Integer>> res, boolean[] visited){
        Queue<Integer> q = new LinkedList<>();
        visited[idx] = true;
        q.add(idx);
        while(!q.isEmpty()){
            int now = q.poll();
            for(int next:map.get(now)){
                if(visited[next]) continue;
                q.add(next);
                res.get(idx).add(next);
                visited[next] = true;
            }
        }
    }

    private void dfs(int idx, int now, Map<Integer, List<Integer>> map, List<List<Integer>> res, boolean[] visited){
        if(visited[now]) return;
        visited[now] = true;
        if(idx!=now) res.get(idx).add(now);
        for(int next:map.get(now)){
            dfs(idx, next, map, res, visited);
        }
        
    }
}

// TC: O(N^3) -> (E*V) = ((N-1)*N)*N
// SC: O(N^2) -> (E+V) = ((N-1)*N)+N

File: leetcode/Medium/2192. All Ancestors of a Node in a Directed Acyclic Graph/2192.cpp
class Solution {
public:
    vector<vector<int>> getAncestors(int n, vector<vector<int>>& edges) {
        vector<vector<int>> res(n);
        unordered_map<int, vector<int>> mp;

        for(auto &e:edges){
            mp[e[1]].push_back(e[0]);
        }

        for(int i=0; i<n; i++){
            vector<bool> visited(n, false);
            bfs(i, mp, res, visited);
            // dfs(i, i, mp, res, visited);
            sort(res[i].begin(), res[i].end());
        }

        return res;
    }

    void bfs(int idx, unordered_map<int, vector<int>>& mp, vector<vector<int>>& res, vector<bool> visited){
        queue<int> q;
        visited[idx] = false;
        q.push(idx);
        while(!q.empty()){
            int now = q.front();
            q.pop();
            for(auto &next:mp[now]){
                if(visited[next]) continue;
                q.push(next);
                res[idx].push_back(next);
                visited[next] = true;
            }
        }
    }

    void dfs(int idx, int now, unordered_map<int, vector<int>>& mp, vector<vector<int>>& res, vector<bool>& visited){
        if(visited[now]) return;
        visited[now] = true;
        if(idx != now) res[idx].push_back(now);
        for(int next: mp[now]){
            dfs(idx, next, mp, res, visited);
        }
    }
};


// TC: O(N^3) -> (EV) = ((N-1)*N) * N
// SC: O(N^2) -> (E+V) = (N-1)*N +N
